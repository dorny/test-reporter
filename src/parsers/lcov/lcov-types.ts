export interface LcovReport {
  [str: string]: {
    path: string
    statementMap: unknown
    fnMap: unknown
    branchMap: unknown
    s: CovStats
    f: CovStats
    b: CovStats
  }
}

export interface CovStats {
  [str: string]: number
}

export interface CovParsedStat {
  max: number
  nonCovered: number
  percentage: number
}
