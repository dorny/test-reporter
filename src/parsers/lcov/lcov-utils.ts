import parse, {LcovFile} from 'lcov-parse'

const parseProm = async (pathOrStr: string): Promise<LcovFile[]> => {
  return new Promise((resolve, reject) => {
    parse(pathOrStr, (err, data) => {
      if (err) {
        reject(err)
      }
      resolve(data ?? [])
    })
  })
}

export {parseProm}
