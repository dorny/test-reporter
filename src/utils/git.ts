export async function listFiles(): Promise<string[]> {
  return []
}
