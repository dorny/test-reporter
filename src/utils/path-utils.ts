export function normalizeDirPath(path: string, addTrailingSlash: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (addTrailingSlash && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}

export function getBasePath(path: string, trackedFiles: string[]): string | undefined {
  if (trackedFiles.includes(path)) {
    return ''
  }

  for (const file of trackedFiles) {
    const pathParts = path.split('/')
    const originalLength = pathParts.length
    const fileParts = file.split('/')

    while (pathParts.length && pathParts.slice(-1)[0] === fileParts.slice(-1)[0]) {
      pathParts.pop()
      fileParts.pop()
    }

    // we found some matching path parts
    if (pathParts.length !== originalLength) {
      return pathParts.join('/')
    }
  }

  return undefined
}
