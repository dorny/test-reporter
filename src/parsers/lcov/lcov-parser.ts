import {ParseOptions, TestParser} from '../../test-parser'

import {TestCaseResult, TestGroupResult, TestRunResult, TestSuiteResult} from '../../test-results'
import {parseProm} from './lcov-utils'
import {LcovBranch, LcovFile, LcovFunc, LcovLine, LcovPart} from 'lcov-parse'

export class LcovParser implements TestParser {
  constructor(readonly options: ParseOptions) {}
  async parse(path: string, content: string): Promise<TestRunResult> {
    const report = await this.parseFile(path, content)
    return this.getTestRunResult(path, report)
  }

  private async parseFile(path: string, content: string): Promise<LcovFile[]> {
    try {
      return parseProm(content)
      //return JSON.parse(content) as LcovReport
    } catch (e) {
      throw new Error(`Invalid JSON at ${path}\n\n${e}`)
    }
  }
  private async getTestRunResult(path: string, report: LcovFile[]): Promise<TestRunResult> {
    const suites: TestSuiteResult[] = []

    for (const reportElement of report) {
      const fileName = reportElement.file

      const statementCaseResult: TestCaseResult = {
        name: `lines ${this.getPartInfo(reportElement.lines)}`,
        time: 0,
        result: this.getPercentage(reportElement.lines) >= 80 ? 'success' : 'failed'
      }
      const fonctionCaseResult: TestCaseResult = {
        name: `functions ${this.getPartInfo(reportElement.functions)}`,
        time: 0,
        result: this.getPercentage(reportElement.functions) >= 80 ? 'success' : 'failed'
      }
      const brancheCaseResult: TestCaseResult = {
        name: `branches ${this.getPartInfo(reportElement.branches)}`,
        time: 0,
        result: this.getPercentage(reportElement.branches) >= 80 ? 'success' : 'failed'
      }

      const testCases: TestCaseResult[] = [statementCaseResult, fonctionCaseResult, brancheCaseResult]
      const groups: TestGroupResult[] = [new TestGroupResult(fileName, testCases)]
      const suite: TestSuiteResult = new TestSuiteResult(fileName, groups)

      suites.push(suite)
    }
    return new TestRunResult(path, suites)
  }

  private getPercentage(stat: LcovPart<LcovLine | LcovFunc | LcovBranch>): number {
    return stat ? (stat.hit / stat.found) * 100 : 100
  }
  private getPartInfo(stat: LcovPart<LcovLine | LcovFunc | LcovBranch>): string {
    return `${this.getPercentage(stat)}% (${stat.hit}/${stat.found})`
  }
}
