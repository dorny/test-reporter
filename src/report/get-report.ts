import * as core from '@actions/core'
import {TestExecutionResult, TestRunResult, TestSuiteResult} from '../test-results'
import {Align, formatTime, Icon, link, table} from '../utils/markdown-utils'
import {DEFAULT_LOCALE} from '../utils/node-utils'
import {getFirstNonEmptyLine} from '../utils/parse-utils'
import {slug} from '../utils/slugger'

const MAX_REPORT_LENGTH = 65535
const MAX_ACTIONS_SUMMARY_LENGTH = 131072 // 1048576 soon

export interface ReportOptions {
  listSuites: 'all' | 'failed' | 'none'
  listTests: 'all' | 'failed' | 'none'
  baseUrl: string
  onlySummary: boolean
  useActionsSummary: boolean
  badgeTitle: string
}

const defaultOptions: ReportOptions = {
  listSuites: 'all',
  listTests: 'all',
  baseUrl: '',
  onlySummary: false,
  useActionsSummary: true,
  badgeTitle: 'tests'
}

export function getReport(results: TestRunResult[], options: ReportOptions = defaultOptions): string {
  core.info('Generating check run summary')

  applySort(results)

  const opts = {...options}
  let lines = renderReport(results, opts)
  let report = lines.join('\n')

  if (getByteLength(report) <= getMaxReportLength(options)) {
    return report
  }

  if (opts.listTests === 'all') {
    core.info("Test report summary is too big - setting 'listTests' to 'failed'")
    opts.listTests = 'failed'
    lines = renderReport(results, opts)
    report = lines.join('\n')
    if (getByteLength(report) <= getMaxReportLength(options)) {
      return report
    }
  }

  core.warning(`Test report summary exceeded limit of ${getMaxReportLength(options)} bytes and will be trimmed`)
  return trimReport(lines, options)
}

function getMaxReportLength(options: ReportOptions = defaultOptions): number {
  return options.useActionsSummary ? MAX_ACTIONS_SUMMARY_LENGTH : MAX_REPORT_LENGTH
}

function trimReport(lines: string[], options: ReportOptions): string {
  const closingBlock = '```'
  const errorMsg = `**Report exceeded GitHub limit of ${getMaxReportLength(options)} bytes and has been trimmed**`
  const maxErrorMsgLength = closingBlock.length + errorMsg.length + 2
  const maxReportLength = getMaxReportLength(options) - maxErrorMsgLength

  let reportLength = 0
  let codeBlock = false
  let endLineIndex = 0
  for (endLineIndex = 0; endLineIndex < lines.length; endLineIndex++) {
    const line = lines[endLineIndex]
    const lineLength = getByteLength(line)

    reportLength += lineLength + 1
    if (reportLength > maxReportLength) {
      break
    }

    if (line === '```') {
      codeBlock = !codeBlock
    }
  }

  const reportLines = lines.slice(0, endLineIndex)
  if (codeBlock) {
    reportLines.push('```')
  }
  reportLines.push(errorMsg)
  return reportLines.join('\n')
}

function applySort(results: TestRunResult[]): void {
  results.sort((a, b) => a.path.localeCompare(b.path, DEFAULT_LOCALE))
  for (const res of results) {
    res.suites.sort((a, b) => a.name.localeCompare(b.name, DEFAULT_LOCALE))
  }
}

function getByteLength(text: string): number {
  return Buffer.byteLength(text, 'utf8')
}

function renderReport(results: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const badge = getReportBadge(results, options)
  sections.push(badge)

  const runs = getTestRunsReport(results, options)
  sections.push(...runs)

  return sections
}

function getReportBadge(results: TestRunResult[], options: ReportOptions): string {
  const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
  const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
  const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
  return getBadge(passed, failed, skipped, options)
}

function getBadge(passed: number, failed: number, skipped: number, options: ReportOptions): string {
  const text = []
  if (passed > 0) {
    text.push(`${passed} passed`)
  }
  if (failed > 0) {
    text.push(`${failed} failed`)
  }
  if (skipped > 0) {
    text.push(`${skipped} skipped`)
  }
  const message = text.length > 0 ? text.join(', ') : 'none'

  let color = 'success'
  if (failed > 0) {
    color = 'critical'
  } else if (passed === 0 && failed === 0) {
    color = 'yellow'
  }
  const hint = failed > 0 ? 'Tests failed' : 'Tests passed successfully'
  const uri = encodeURIComponent(`${options.badgeTitle}-${message}-${color}`)
  return `![${hint}](https://img.shields.io/badge/${uri})`
}

function getTestRunsReport(testRuns: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const totalFailed = testRuns.reduce((sum, tr) => sum + tr.failed, 0)
  if (totalFailed === 0) {
    sections.push(`<details><summary>Expand for details</summary>`)
    sections.push(` `)
  }

  const shouldShowCoverage = testRuns.some(tr => tr.coverage !== undefined)

  const columnNames = ['Report', 'Passed', 'Failed', 'Skipped', 'Time']
  const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

  if (shouldShowCoverage) {
    columnNames.push('Coverage')
    columnAligns.push(Align.Right)
  }

  if (testRuns.length > 0 || options.onlySummary) {
    const tableData = testRuns
      .filter(tr => tr.passed > 0 || tr.failed > 0 || tr.skipped > 0)
      .map(tr => {
        const time = formatTime(tr.time)
        const name = tr.path
        const passed = tr.passed > 0 ? `${tr.passed} ${Icon.success}` : ''
        const failed = tr.failed > 0 ? `${tr.failed} ${Icon.fail}` : ''
        const skipped = tr.skipped > 0 ? `${tr.skipped} ${Icon.skip}` : ''

        if (shouldShowCoverage) {
          const coverage = tr.coverage !== undefined ? tr.coverage + '%' : ''
          return [name, passed, failed, skipped, time, coverage]
        }

        return [name, passed, failed, skipped, time]
      })

    const resultsTable = table(columnNames, columnAligns, ...tableData)
    sections.push(resultsTable)
  }

  if (options.onlySummary === false) {
    const suitesReports = testRuns.map((tr, i) => getSuitesReport(tr, i, options)).flat()
    sections.push(...suitesReports)
  }

  if (totalFailed === 0) {
    sections.push(`</details>`)
  }
  return sections
}

function getSuitesReport(tr: TestRunResult, runIndex: number, options: ReportOptions): string[] {
  const sections: string[] = []
  const suites = options.listSuites === 'failed' ? tr.failedSuites : tr.suites

  if (options.listSuites !== 'none') {
    const trSlug = makeRunSlug(runIndex)
    const nameLink = `<a id="${trSlug.id}" href="${options.baseUrl + trSlug.link}">${tr.path}</a>`
    const icon = getResultIcon(tr.result)
    sections.push(`## ${icon}\xa0${nameLink}`)

    const time = formatTime(tr.time)
    const headingLine2 =
      tr.tests > 0
        ? `**${tr.tests}** tests were completed in **${time}** with **${tr.passed}** passed, **${tr.failed}** failed and **${tr.skipped}** skipped.`
        : 'No tests found'
    sections.push(headingLine2)

    if (suites.length > 0) {
      const shouldShowCoverage = suites.some(s => s.coverage !== undefined)

      const columnNames = ['Test suite', 'Passed', 'Failed', 'Skipped', 'Time']
      const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

      if (shouldShowCoverage) {
        columnNames.push('Coverage')
        columnAligns.push(Align.Right)
      }

      const suitesTable = table(
        columnNames,
        columnAligns,
        ...suites.map((s, suiteIndex) => {
          const tsTime = formatTime(s.time)
          const tsName = s.name
          const skipLink = options.listTests === 'none' || (options.listTests === 'failed' && s.result !== 'failed')
          const tsAddr = options.baseUrl + makeSuiteSlug(runIndex, suiteIndex).link
          const tsNameLink = skipLink ? tsName : link(tsName, tsAddr)
          const passed = s.passed > 0 ? `${s.passed} ${Icon.success}` : ''
          const failed = s.failed > 0 ? `${s.failed} ${Icon.fail}` : ''
          const skipped = s.skipped > 0 ? `${s.skipped} ${Icon.skip}` : ''

          const coverage = s.coverage !== undefined ? s.coverage + '%' : ''

          if (shouldShowCoverage) {
            return [tsNameLink, passed, failed, skipped, tsTime, coverage]
          }

          return [tsNameLink, passed, failed, skipped, tsTime]
        })
      )
      sections.push(suitesTable)
    }
  }

  if (options.listTests !== 'none') {
    const tests = suites.map((ts, suiteIndex) => getTestsReport(ts, runIndex, suiteIndex, options)).flat()

    if (tests.length > 1) {
      sections.push(...tests)
    }
  }

  return sections
}

function getTestsReport(ts: TestSuiteResult, runIndex: number, suiteIndex: number, options: ReportOptions): string[] {
  if (options.listTests === 'failed' && ts.result !== 'failed') {
    return []
  }
  const groups = ts.groups
  if (groups.length === 0) {
    return []
  }

  const sections: string[] = []

  const tsName = ts.name
  const tsSlug = makeSuiteSlug(runIndex, suiteIndex)
  const tsNameLink = `<a id="${tsSlug.id}" href="${options.baseUrl + tsSlug.link}">${tsName}</a>`
  const icon = getResultIcon(ts.result)
  sections.push(`### ${icon}\xa0${tsNameLink}`)

  sections.push('```')
  for (const grp of groups) {
    if (grp.name) {
      sections.push(grp.name)
    }
    const space = grp.name ? '  ' : ''
    for (const tc of grp.tests) {
      const result = getResultIcon(tc.result)
      sections.push(`${space}${result} ${tc.name}`)
      if (tc.error) {
        const lines = (tc.error.message ?? getFirstNonEmptyLine(tc.error.details)?.trim())
          ?.split(/\r?\n/g)
          .map(l => '\t' + l)
        if (lines) {
          sections.push(...lines)
        }
      }
    }
  }
  sections.push('```')

  return sections
}

function makeRunSlug(runIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}`)
}

function makeSuiteSlug(runIndex: number, suiteIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}s${suiteIndex}`)
}

function getResultIcon(result: TestExecutionResult): string {
  switch (result) {
    case 'success':
      return Icon.success
    case 'skipped':
      return Icon.skip
    case 'failed':
      return Icon.fail
    default:
      return ''
  }
}
