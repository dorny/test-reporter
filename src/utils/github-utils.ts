import {createWriteStream} from 'fs'
import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'
import * as stream from 'stream'
import {promisify} from 'util'
import got from 'got'
const asyncStream = promisify(stream.pipeline)

export function getCheckRunContext(): {sha: string; runId: number; branch: string} {
  let branch = github.context.ref
  if (branch.startsWith('refs/heads')) branch = branch.slice(11)
  core.info('Branch: ' + branch)

  if (github.context.eventName === 'workflow_run') {
    core.info('Action was triggered by workflow_run: using SHA and RUN_ID from triggering workflow')
    const event = github.context.payload
    if (!event.workflow_run) {
      throw new Error("Event of type 'workflow_run' is missing 'workflow_run' field")
    }
    return {
      sha: event.workflow_run.head_commit.id,
      runId: event.workflow_run.id,
      branch
    }
  }

  const runId = github.context.runId
  if (github.context.payload.pull_request) {
    core.info(`Action was triggered by ${github.context.eventName}: using SHA from head of source branch`)
    const pr = github.context.payload.pull_request
    return {sha: pr.head.sha, runId, branch}
  }

  return {sha: github.context.sha, runId, branch}
}

export async function downloadArtifact(
  octokit: InstanceType<typeof GitHub>,
  artifactId: number,
  fileName: string,
  token: string
): Promise<void> {
  core.startGroup(`Downloading artifact ${fileName}`)
  try {
    core.info(`Artifact ID: ${artifactId}`)

    const req = octokit.rest.actions.downloadArtifact.endpoint({
      ...github.context.repo,
      artifact_id: artifactId,
      archive_format: 'zip'
    })

    const headers = {
      Authorization: `Bearer ${token}`
    }
    const resp = await got(req.url, {
      headers,
      followRedirect: false
    })

    core.info(`Fetch artifact URL: ${resp.statusCode} ${resp.statusMessage}`)
    if (resp.statusCode !== 302) {
      throw new Error('Fetch artifact URL failed: received unexpected status code')
    }

    const url = resp.headers.location
    if (url === undefined) {
      const receivedHeaders = Object.keys(resp.headers)
      core.info(`Received headers: ${receivedHeaders.join(', ')}`)
      throw new Error('Location header was not found in API response')
    }
    if (typeof url !== 'string') {
      throw new Error(`Location header has unexpected value: ${url}`)
    }

    const downloadStream = got.stream(url, {headers})
    const fileWriterStream = createWriteStream(fileName)

    core.info(`Downloading ${url}`)
    downloadStream.on('downloadProgress', ({transferred}) => {
      core.info(`Progress: ${transferred} B`)
    })
    await asyncStream(downloadStream, fileWriterStream)
  } finally {
    core.endGroup()
  }
}

export async function listFiles(octokit: InstanceType<typeof GitHub>, sha: string): Promise<string[]> {
  core.startGroup('Fetching list of tracked files from GitHub')
  try {
    const commit = await octokit.rest.git.getCommit({
      commit_sha: sha,
      ...github.context.repo
    })
    const files = await listGitTree(octokit, commit.data.tree.sha, '')
    return files
  } finally {
    core.endGroup()
  }
}

async function listGitTree(octokit: InstanceType<typeof GitHub>, sha: string, path: string): Promise<string[]> {
  const pathLog = path ? ` at ${path}` : ''
  core.info(`Fetching tree ${sha}${pathLog}`)
  let truncated = false
  let tree = await octokit.rest.git.getTree({
    recursive: 'true',
    tree_sha: sha,
    ...github.context.repo
  })

  if (tree.data.truncated) {
    truncated = true
    tree = await octokit.rest.git.getTree({
      tree_sha: sha,
      ...github.context.repo
    })
  }

  const result: string[] = []
  for (const tr of tree.data.tree) {
    const file = `${path}${tr.path}`
    if (tr.type === 'blob') {
      result.push(file)
    } else if (tr.type === 'tree' && truncated) {
      const files = await listGitTree(octokit, tr.sha || '', `${file}/`)
      result.push(...files)
    }
  }

  return result
}
