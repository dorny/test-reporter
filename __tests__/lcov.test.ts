import * as fs from 'fs'
import * as path from 'path'

import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'
import {LcovParser} from '../src/parsers/lcov/lcov-parser'

describe('lcov report coverage', () => {
  it('report from facebook/jest test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'lcov.info')
    const outputPath = path.join(__dirname, '__outputs__', 'lcov-report-results.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const parser = new LcovParser({parseErrors: true, trackedFiles: []})
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
