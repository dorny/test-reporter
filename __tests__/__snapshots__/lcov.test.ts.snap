// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lcov report coverage report from facebook/jest test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/lcov.info",
  "suites": [
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": "src/services/notifier/NotifierService.js",
          "tests": [
            {
              "name": "lines 100% (21/21)",
              "result": "success",
              "time": 0,
            },
            {
              "name": "functions 100% (10/10)",
              "result": "success",
              "time": 0,
            },
            {
              "name": "branches 50% (3/6)",
              "result": "failed",
              "time": 0,
            },
          ],
        },
      ],
      "name": "src/services/notifier/NotifierService.js",
      "totalTime": undefined,
    },
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": "src/services/notifier/providers/DiscordNotifierProvider.js",
          "tests": [
            {
              "name": "lines 100% (17/17)",
              "result": "success",
              "time": 0,
            },
            {
              "name": "functions 100% (3/3)",
              "result": "success",
              "time": 0,
            },
            {
              "name": "branches 75% (3/4)",
              "result": "failed",
              "time": 0,
            },
          ],
        },
      ],
      "name": "src/services/notifier/providers/DiscordNotifierProvider.js",
      "totalTime": undefined,
    },
  ],
  "totalTime": undefined,
}
`;
